src/ray/object_manager/object_manager.cc

```diff
@@ -412,47 +417,58 @@ ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t
 void ObjectManager::AllWaitLookupsComplete(const UniqueID &wait_id) {
   auto &wait_state = active_wait_requests_.find(wait_id)->second;
   if (wait_state.found.size() >= wait_state.num_required_objects ||
-      wait_state.wait_ms == 0) {
+      wait_state.timeout_ms == 0) {
     // Requirements already satisfied.
     WaitComplete(wait_id);
   } else {
-    for (auto &oid : wait_state.remaining) {
+    // Subscribe to objects in order to ensure Wait-related tests are deterministic.
+    for (auto &object_id : wait_state.object_id_order) {
+      if (wait_state.remaining.count(object_id) == 0) {
```

Comment:
This is the relevant check that corrects the bug you described earlier: `active_wait_requests_.find(wait_id) == active_wait_requests_.end()`

Reply:
Yes, I don't think it will break in that particular way anymore, but undefined behavior is still possible because of the reference to `wait_state`. Same underlying issue, but it will break at a different line.

Reply:
Nice catch. I've made the fix along with a regression test.